File: src/Promote.java
public class Promote implements Event{
    private final Position at;

    public Promote(Position at) {
        this.at = at;
    }

    @Override
    public void doToBoard(Board board) {
        board.replaceAtPosition(at, new Queen(board.atPosition(at).color));
    }

    @Override
    public void undoToBoard(Board board) {
        board.replaceAtPosition(at, new Pawn(board.atPosition(at).color));
    }
}

File: src/Promote.kt
class Promote(private val at: Position) : Event {
    override fun doToBoard(board: Board) {
        board.replaceAtPosition(at, Queen(board.atPosition(at).color))
    }

    override fun undoToBoard(board: Board) {
        board.replaceAtPosition(at, Pawn(board.atPosition(at).color))
    }
}